static/js/tools/map/chart_store.ts

```diff
@@ -0,0 +1,199 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import _ from "lodash";
+
+import { GeoJsonData, MapPoint } from "../../chart/types";
+import {
+  EntityObservationListWrapper,
+  EntityObservationWrapper,
+  EntitySeriesWrapper,
+} from "../../shared/stat_types";
+import { DataContext, PlaceInfo, StatVar } from "./context";
+
+export enum ChartDataType {
+  geoJson,
+  defaultStat,
+  allStat,
+  denomStat,
+  breadcrumbStat,
+  breadcrumbDenomStat,
+  mapPointStat,
+  mapPointCoordinate,
+}
+
+// ChartStore holds the raw data and corresponding context.
+// When context changes, the data fetch is async and could take long time. The
+// context here is used to check if the data context matches the actual context.
+export interface ChartStore {
+  geoJson: {
+    data: GeoJsonData;
+    context?: DataContext;
+  };
+  defaultStat: {
+    data: EntityObservationWrapper;
+    context?: DataContext;
+  };
+  allStat: {
+    data: EntityObservationListWrapper;
+    context?: DataContext;
+  };
+  denomStat: {
+    data: EntitySeriesWrapper;
+    context?: DataContext;
+  };
+  breadcrumbStat: {
+    data: EntityObservationWrapper;
+    context?: DataContext;
+  };
+  breadcrumbDenomStat: {
+    data: EntitySeriesWrapper;
+    context?: DataContext;
+  };
+  mapPointStat: {
+    data: EntityObservationWrapper;
+    context?: DataContext;
+  };
+  mapPointCoordinates: {
+    data: Array<MapPoint>;
+    context?: DataContext;
+  };
+}
+
+export const emptyChartStore = {
+  geoJson: {
+    data: null,
+  },
+  defaultStat: {
+    data: null,
+  },
+  allStat: {
+    data: null,
+  },
+  denomStat: {
+    data: null,
+  },
+  breadcrumbStat: {
+    data: null,
+  },
+  breadcrumbDenomStat: {
+    data: null,
+  },
+  mapPointStat: {
+    data: null,
+  },
+  mapPointCoordinates: {
+    data: null,
+  },
+};
+
+/**
+ * Check whether data is ready to use.
+ *
+ * Data is ready if the context in ChartStore matches the current context.
+ *
+ * @param ctx The data context from chart store.
+ * @param type The type of the data.
+ * @param placeInfo The current placeInfo from the context.
+ * @param statVar The current statVar from the context.
+ * @returns
+ */
+export const isDataReady = (
+  dataWrapper: { data: any; context?: DataContext },
+  type: ChartDataType,
+  placeInfo: PlaceInfo,
+  statVar: StatVar
+): boolean => {
+  if (_.isEmpty(dataWrapper)) {
+    return false;
+  }
+  const ctx = dataWrapper.context;
+  if (_.isEmpty(ctx)) {
+    return false;
+  }
+  switch (type) {
```

Comment:
nit: I feel like this is a large switch statement with lots of checks for each case making it not the easiest to read. Maybe we can extract out functions for each case and just call the function for each case? or even make a map of type to function & just do something like `return TYPE_TO_READY_FN[type]`

Reply:
I later realized it's better to move this as useCallback() function inside <ChartLoader> and only use the necessary dependent variables instead of StatVar and PlaceInfo struct. That way, the function return value changes only when those variables change, which can eliminate unnecessary re-renders.

With that, I split this into separate useCallback functions.